intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp

```diff
@@ -0,0 +1,487 @@
+
+      // original 2
+      // value_layer = ne_view_3d(gctx, v_cache, n_past + qlen, head_size, num_attention_heads, v_cache->nb[1],v_cache->nb[2], 0); // [heads, head_size, klen]
+
+      // attention
+      struct ne_tensor *attn_scores = ne_mul_mat(ctx0, key_layer, query_layer); // [kv_heads, mqa_scale * qlen, klen]
+      ne_set_name(attn_scores, "attn_scores");
+
+      // 2
+      // attn_scores = ne_scale_inplace(ctx0, attn_scores, ne_new_f32(ctx0, 1.f / std::sqrt(head_size)));
+
+      // if (n_past == 0) {
+      //   // build attention mask for context input
+      //   attn_scores = ne_reshape_3d(ctx0, attn_scores, n_past + qlen, qlen,
+      //                                 num_attention_heads); // [heads, qlen, klen]
+      //   attn_scores = ne_diag_mask_inf_inplace(ctx0, attn_scores, n_past);
+      //   attn_scores = ne_reshape_3d(ctx0, attn_scores, n_past + qlen, mqa_scale * qlen,
+      //                                 num_kv_heads); // [kv_heads, mqa_scale * qlen, klen]
+      // }
+
+      if (n_past == 0) {
+        // build attention mask for context input
+        ne_tensor *inf = ne_new_tensor_3d(ctx0, attn_scores->type, 1, qlen - 1, num_attention_heads, NE_SIZE_CALC);
+        ne_set_f32(inf, -INFINITY);
+
+        ne_tensor *masked_attn_scores =
+            ne_view_3d(ctx0, attn_scores, 1, qlen - 1, num_attention_heads, qlen * ne_element_size(attn_scores),
+                         qlen * qlen * ne_element_size(attn_scores), (qlen - 1) * ne_element_size(attn_scores));
+
+        ne_set_name(masked_attn_scores, "masked_attn_scores");
+        ne_build_forward_expand(&gf, ne_cpy(ctx0, inf, masked_attn_scores));
+      }
+
+      // struct ne_tensor *attn_probs = ne_soft_max_inplace(ctx0, attn_scores); // [kv_heads, mqa_scale * qlen, klen]
+
+      // struct ne_tensor *context_layer = ne_mul_mat(ctx0, value_layer, attn_probs); // [kv_heads, mqa_scale * qlen, head_size]
+      // context_layer = ne_reshape_3d(ctx0, context_layer, head_size, qlen,
+      //                                 num_attention_heads);                           // [heads, qlen, head_size]
+      // context_layer = ne_cont(ctx0, ne_permute(ctx0, context_layer, 0, 2, 1, 3)); // [qlen, heads, head_size]
+      // context_layer = ne_reshape_2d(ctx0, context_layer, hidden_size, qlen); // [qlen, hidden]
+
+      attn_scores = ne_scale_inplace(ctx0, attn_scores, ne_new_f32(ctx0, 1.f / std::sqrt(head_size)));
+      
+      ne_tensor *attn_probs = ne_soft_max_inplace(ctx0, attn_scores); // [heads, qlen, klen]
+
+      ne_tensor *context_layer = ne_mul_mat(ctx0, value_layer, attn_probs); // [heads, qlen, head_size]
+      
+      context_layer = ne_cont(ctx0, ne_permute(ctx0, context_layer, 0, 2, 1, 3));
+      
+      context_layer = ne_reshape_2d(ctx0, context_layer, hidden_size, qlen);
+
+      // struct ne_tensor *attn_output = dense.forward(ctx0, context_layer);
+      // Linear如果有bias就需要add，如果没有就不需要。到底有没有bias呢？？？？？？
```

Comment:
it depends on model arch. del this annotation？

Reply:
Yes, chatglm-1 and 2 are different. I will do cleancode first.

Reply:
fixed